semantic-analysis/src/Analysis/Typecheck.hs

```diff
@@ -21,154 +22,144 @@ module Analysis.Typecheck
 , typecheckingFlowInsensitive
 ) where
 
-import           Analysis.Carrier.Env.Monovariant
-import qualified Analysis.Carrier.Heap.Monovariant as A
-import qualified Analysis.Effect.Domain as A
+import           Analysis.Carrier.Fail.WithLoc
+import qualified Analysis.Carrier.Store.Monovariant as A
+import           Analysis.Effect.Domain as A
 import           Analysis.File
 import           Analysis.FlowInsensitive
 import           Analysis.Functor.Named
-import qualified Analysis.Intro as Intro
+import           Analysis.Reference
 import           Control.Algebra
-import           Control.Applicative (Alternative (..))
-import           Control.Carrier.Fail.WithLoc
+import           Control.Applicative (Alternative(..))
 import           Control.Carrier.Fresh.Strict as Fresh
 import           Control.Carrier.Reader hiding (Local)
 import           Control.Carrier.State.Strict
-import           Control.Monad (unless)
+import           Control.Effect.Labelled
+import           Control.Monad (ap, guard, unless)
 import           Control.Monad.Trans.Class
 import           Data.Foldable (for_)
 import           Data.Function (fix)
-import           Data.Functor (($>))
-import qualified Data.IntMap as IntMap
-import qualified Data.IntSet as IntSet
-import qualified Data.Map as Map
+import qualified Data.IntMap as IM
+import qualified Data.IntSet as IS
 import           Data.Maybe (fromJust, fromMaybe)
-import           Data.Semigroup (Last (..))
+import           Data.Semigroup (Last(..))
 import qualified Data.Set as Set
-import           Data.Traversable (for)
 import           Data.Void
 import           GHC.Generics (Generic1)
 import           Prelude hiding (fail)
-import           Source.Span
-import qualified Syntax.Algebra as Syntax
-import           Syntax.Functor
-import           Syntax.Module
-import           Syntax.Scope
-import qualified Syntax.Sum as Syntax
-import           Syntax.Term
-import           Syntax.Var (closed)
-import qualified System.Path as Path
-
-data Monotype f a
-  = Bool
+
+data Monotype a
+  = Var a
   | Unit
+  | Bool
+  | Int
   | String
-  | f a :-> f a
-  | Record (Map.Map Name (f a))
-  deriving (Foldable, Functor, Generic1, Traversable)
+  | Monotype a :-> Monotype a
```

Comment:
We no longer represent polytypes in quite the same way, but rather as a right `Type`-module, so we can recur directly here and just wrap the whole thing in some foralls binding type variables at the end. That’s actually more in line with Hindley-Damas-Milner anyway, as it’s essentially type schemes.